### ISSUE TITLE
Using LLMs to generate semantic topic labels

### ISSUE THREAD

**Issue by clstaudt** [MaartenGr/BERTopic#1534]
I passed the topic representations (lists of keywords) to GPT-4 to generate a topic label that represents the topic. At first glance the results are promising and could in some cases replace expert labelling of the topics. 

Could this maybe become a built-in feature of BERTopic?

![Screenshot 2023-09-20 at 10 31 29](https://github.com/MaartenGr/BERTopic/assets/875194/d9f6fd0f-1fec-45af-92e1-409801dfee7c)

```python
def generate_topic_labels(dataframe, input_column, output_column):
    """
    Generate topic labels for a given DataFrame column using the OpenAI API.
    
    Parameters:
    - dataframe (pd.DataFrame): The input DataFrame.
    - input_column (str): The name of the column in the DataFrame containing keyword lists.
    - output_column (str): The name of the new column to be added to the DataFrame containing the generated topic labels.
    
    Returns:
    - pd.DataFrame: The DataFrame with an additional column of topic labels.
    """
    def get_topic(keywords):
        # Use OpenAI API to generate a response
        response = openai.ChatCompletion.create(
            model="gpt-4",
            messages=[
                {"role": "system", "content": "You are an AI assistant well versed in topic modelling on documents. If you are unable to perform the task, output: None"},
                {
                    "role": "user", 
                    "content": f"""
                    Output a concise, English, lowercase topic label for the following keywords. Output only the label. Prefer single terms. 
                    {', '.join(keywords)}
                    """},
            ]
        )
        
        # Extract the topic label from the response
        topic_label = response['choices'][0]['message']['content']
        
        return topic_label
    
    # Apply the function to the dataframe
    dataframe[output_column] = dataframe[input_column].apply(get_topic)

    return dataframe
``

**Comment by MaartenGr** [comment]
It already is! You can check the following for more information: https://maartengr.github.io/BERTopic/getting_started/representation/llm.html

**Comment by clstaudt** [comment]
I tried to replace the above code with the built-in solution:


```python
representation_model = OpenAI(
    model="gpt-4", 
    delay_in_seconds=10, 
    chat=True,
    prompt=f"""
        Output a concise, English, lowercase topic label for the following keywords. Output only the label. Prefer single terms. If you are unable to perform the task, output: None. 
        [KEYWORDS]
    """
)

topic_model_custom_prompt = BERTopic(
    embedding_model=embedding_model,
    representation_model=representation_model,
    nr_topics="auto",  # To automatically reduce the number of topics after extraction
    #low_memory=True,  # Useful for large datasets
)

topics, probs = topic_model_custom_prompt.fit_transform(docs)
```

I am surprised how long the call to fit_transform is taking now: 4x the running time of the custom function above so far (12 minutes) and it has not finished yet.

**Comment by MaartenGr** [comment]
This might relate to the `nr_topics="auto"` parameter which will generate the representations again after doing the initial representations. Generally, and according to the [best practices](https://maartengr.github.io/BERTopic/getting_started/best_practices/best_practices.html), I would advise not using that parameter but controlling the number of topics through `min_topic_size`. 

Also, I see that you set `delay_in_seconds=10` which means that between each OpenAI API call there is a 10 second wait which will indeed increase the time.

**Comment by clstaudt** [comment]
Indeed, it's the delay, which can just be removed, no rate limit issues for me so far.


Now I am getting the expected results in the expected time:

![Screenshot 2023-09-20 at 14 39 29](https://github.com/MaartenGr/BERTopic/assets/875194/3b9ca521-1b53-4946-9d40-c35b9b10e44c)

However, I would like to preserve the original list of keywords representation in the topic_info. Is that possible?

**Comment by MaartenGr** [comment]
Yes, that is possible if you enable [multiple representations](https://maartengr.github.io/BERTopic/getting_started/multiaspect/multiaspect.html)

**Comment by clstaudt** [comment]
How do I define the default representation as one Aspect?

**Comment by clstaudt** [comment]
Thank you @MaartenGr
